futures-02.md

```diff
@@ -0,0 +1,396 @@
+          where F: FnOnce(&mut T) -> R
+      {
+          ...
+      }
+  }
+
+  /// A handle used to awaken a task.
+  pub struct Waker {
+
+      /// Wakes up the associated task.
+      ///
+      /// This signals that the futures associated with the task are ready to
+      /// be `poll`ed again.
+      pub fn wake(&self) { ... }
+  }
+
+  ...
+}
+```
+
+The modified traits will look as follows:
+
+```rust
+pub trait Future {
+    type Item;
+    type Error;
+    fn poll(&mut self, ctx: &mut Context) -> Poll<Self::Item, Self::Error>;
+}
+
+pub trait Stream {
+    type Item;
+    type Error;
+    fn poll_next(&mut self, ctx: &mut Context) -> Poll<Option<Self::Item>, Self::Error>;
+}
+
+pub trait AsyncRead {
+    type Error;
+    unsafe fn initializer(&self) -> Initializer { ... }
+
+    fn poll_read(&mut self, buf: &mut [u8], ctx: &mut Context)
+        -> Poll<usize, Self::Error>;
+
+    fn poll_vectored_read(&mut self, vec: &mut [&mut IoVec], ctx: &mut Context)
+        -> Poll<usize, Self::Error> { ... }
+}
+
+pub trait AsyncWrite {
+    type Error;
+    fn poll_write(&mut self, buf: &[u8], ctx: &mut Context)
+        -> Poll<usize, Self::Error>;
+
+    fn poll_vectored_write(&mut self, vec: &[&IoVec], ctx: &mut Context)
+        -> Poll<usize, Self::Error> { ... }
+
+    fn poll_close(&mut self, ctx: &mut Context) -> Poll<(), Self::Error>;
+}
+```
+
+In terms of the old `futures` task model, the `Waker` is analogous to the `Task`
+returned by [`task::current`][current], and the `wake` method is analogous to
+the old [`Task::notify`][notify].
+
+Astute readers will notice other changes in the traits shown above.
+These will be addressed in later sections.
+
+[current]: https://docs.rs/futures/0.1.16/futures/task/fn.current.html
+[notify]: https://docs.rs/futures/0.1.16/futures/task/struct.Task.html#method.notify
+
+## Crate Separation
+
+The `futures` crate will be split into several different libraries.
+
+### `futures-core`
+
+`futures-core` will include `Future`, `Stream`, `IntoFuture`
+traits and their dependencies. This is a central crate that includes only
+essential items, designed to minimize future breakage. Combinator methods
+will be removed from the `Future` and `Stream` and moved to the
+`FutureExt` and `StreamExt` traits in the `futures-util` crate. `poll` 
+will be the only remaining method on the `Future` and `Stream` traits.
+This separation allows combinator functions to evolve over time without
+breaking changes to the public API surface of async libraries.
+
+### `futures-io`
+
+`futures-io` includes the updated `AsyncRead` and `AsyncWrite` traits.
+These traits were previously included in the `tokio-io` crate.
+However, `AsyncRead` and `AsyncWrite` don't have any dependencies on the `tokio`
+crate or the rest of the `tokio` stack, so they will be moved into the `futures`
+group of libraries. `futures-io` depends only on the `futures-core` crate
+and the `bytes` crate (for the [`Buf`] and [`BufMut`] traits).
```

Comment:
Further down, it says the `bytes` dependency is removed. Is this out-of-date?

If so, why the removal of `Buf`/`BufMut`? They are extemely useful abstractions, and greatly ease using vector IO.

Reply:
Yes, this was out-of-date-- `bytes` is no longer depended upon. You're still more than welcome to use `Buf`/`BufMut`! I agree that they're useful abstractions. However, baking them into the `futures-io` traits means that there is a larger body of code (the `bytes` crate) which has to stabilize before `futures-io` can reach 1.0.